Core/include/Acts/Geometry/Portal.hpp

```diff
@@ -0,0 +1,234 @@
+  }
+};
+
+/// A portal connects two or more neighboring volumes. Each volume has a set of
+/// portals that describes which volumes lie behind the portal in that
+/// direction. Portals use associated portal links to perform lookups of target
+/// volumes.
+/// Each portal has two links, and a corresponding surface. One link is
+/// associated with the direction along the surface's normal vector, and one
+/// with the opposite direction.
+class Portal {
+ public:
+  /// Constructor for a portal from a single link
+  /// @param direction The direction of the link
+  /// @param link The portal link
+  Portal(Direction direction, std::unique_ptr<PortalLinkBase> link);
+
+  /// Constructor for a portal from a surface and volume, where a trivial portal
+  /// link is automatically constructed.
+  /// @param direction The direction of the link
+  /// @param surface The surface from which to create the portal link
+  /// @param volume The volume this portal connects to in the @p direction
+  ///               relative to the normal of @p surface.
+  Portal(Direction direction, std::shared_ptr<RegularSurface> surface,
+         TrackingVolume& volume);
+
+  /// Constructor for a portal from two links. One of the links can be
+  /// `nullptr`, but at least one of them needs to be set. If both are set, they
+  /// need to be valid compatible links that can be fused.
+  /// @param gctx The geometry context
+  /// @param alongNormal The link along the normal of the surface
+  /// @param oppositeNormal The link opposite to the normal of the
+  Portal(const GeometryContext& gctx,
+         std::unique_ptr<PortalLinkBase> alongNormal,
+         std::unique_ptr<PortalLinkBase> oppositeNormal);
+
+  /// Helper struct for the arguments to the portal constructor below using
+  /// designated initializers.
+  struct Config {
+    /// Aggregate over a surface and a volume with optional semantics
+    struct Link {
+      Link() = default;
+      /// Constructor from a surface and a volume
+      Link(std::shared_ptr<RegularSurface> surface, TrackingVolume& volume)
+          : m_surface(std::move(surface)), m_volume(&volume) {}
+
+      /// The associated surface
+      std::shared_ptr<RegularSurface> m_surface = nullptr;
+      /// The associated volume
+      TrackingVolume* m_volume = nullptr;
+    };
+
+    /// Entry for the link along normal
+    /// Entry for the link opposite normal
+    Link alongNormal{};
+    Link oppositeNormal{};
+  };
+
+  /// Constructor that takes a geometry context and an rvalue reference to a
+  /// helper config struct from above. This pattern allows you to use designated
+  /// initializers to construct this object like:
+  /// ```cpp
+  /// Portal{gctx, {.oppositeNormal = {cyl1, *vol1}}};
+  /// Portal{gctx, {.alongNormal = {cyl2, *vol2}}};
+  /// ```
+  /// @param gctx The geometry context
+  /// @param config The configuration struct
+  Portal(const GeometryContext& gctx, Config&& config);
```

Comment:
Alternatively we could have a boolean and/or factory methods to express the different construction ways explicitly

Reply:
This is definitely expressible in a different way, like for example the other constructor which takes two finished links. I was just thinking that this mechanism here, where you have two optional `kwarg` style arguments looks *nice* with designated initializers.

Reply:
This designated initializer thing should also work with normal const lvalue reference instead of rvalue referance, is there a specific reason you chose rvalue reference here?

Reply:
My idea was to use the `struct` only for arguments to this function, `kwargs`-style. The rvalue ref sort of encourages this I think, but I can also change it of course.

Reply:
Ah I see... I mean one can of course always `std::move` the struct into the function if one does'nt like the kwargs-style...
But on the other hand its kind of our convention to have the config struct with const lvalue ref. I definitvely like the `kwargs` approach, but I'm not sure if we should break the convention for that... But up to you.

Reply:
I see - to me taking two `Link` params seems a bit simpler and does not trade much in "expressiveness" for modern editors. But I have no strong opinion for or against this

Reply:
Renamed `Config` here to `Arguments` to underline the purpose.